app/models/samples_report.rb

```diff
@@ -27,27 +29,58 @@ def self.entity_scope
-  }
+    joins("LEFT JOIN samples_report_samples ON samples_report_samples.samples_report_id = samples_reports.id
+      LEFT JOIN samples ON samples_report_samples.sample_id = samples.id
+      LEFT JOIN boxes ON boxes.id = samples.box_id")
+      .where("boxes.uuid LIKE ?", "%#{sanitize_sql_like(box_uuid)}%")
+      .group(:samples_report_id) unless box_uuid.blank?
+    }
 
-  scope :partial_batch_number, ->(batch_number) {
-    joins("LEFT JOIN samples_report_samples ON samples_report_samples.samples_report_id = samples_reports.id 
-            LEFT JOIN samples ON samples_report_samples.sample_id = samples.id 
-            LEFT JOIN batches ON batches.id = samples.batch_id")
+    scope :partial_batch_number, ->(batch_number) {
+      joins("LEFT JOIN samples_report_samples ON samples_report_samples.samples_report_id = samples_reports.id
+        LEFT JOIN samples ON samples_report_samples.sample_id = samples.id
+        LEFT JOIN batches ON batches.id = samples.batch_id")
         .where("batches.batch_number LIKE ?", "%#{sanitize_sql_like(batch_number)}%")
         .group(:samples_report_id) unless batch_number.blank?
-  }
+      }
 
+      def calculate_lod_and_lob
+        concentrations = samples.map(&:concentration)
+        signals = Numo::DFloat[*samples.map(&:measured_signal)]
+        self.lod, _, self.lob, _ = calculate_lod(concentrations, signals)
+        self.save
+      end
 
-  private
 
-  def there_are_samples
-    samples_with_measurements = samples_report_samples.select { |srs| !srs.sample.measured_signal.nil? }
-    errors.add(:base, "The selected box should contain samples with uploaded measurements") if samples_with_measurements.empty?
-  end
+      private
+
+      def there_are_samples
+        samples_with_measurements = samples_report_samples.select { |srs| !srs.sample.measured_signal.nil? }
+        errors.add(:base, "The selected box should contain samples with uploaded measurements") if samples_with_measurements.empty?
+      end
+
+      def calculate_lod(concentrations, signals)
+        concentrations_matrix = Numo::DFloat[*concentrations.map { |c| [c] }]
+        lr = Rumale::LinearModel::LinearRegression.new(fit_bias: true)
+        lr.fit(concentrations_matrix, signals)
+
+        lod = lr.bias_term
+        predictions = lr.predict(concentrations_matrix)
+        residuals = signals - predictions
+        rmse = Math.sqrt((residuals**2).sum / residuals.size)
+
+        blank_samples = signals[Numo::DFloat.cast(concentrations).eq(0)]
+        if blank_samples.size > 0
+          blank_samples_mean = blank_samples.mean
+          blank_samples_sd = blank_samples.stddev
+          lod = [lod, 3 * blank_samples_sd].max
+          lob = blank_samples_mean + 1.645 * blank_samples_sd
+        else
+          lob = nil
+        end
+
+        gray_zone = lob.nil? ? nil : (lob..lod)
+        return lod.round(3), rmse, lob.round(3), gray_zone
+      end
 
-end
\ No newline at end of file
+    end
```

Comment:
nitpick: indentation looks wrong in the whole file? it seems it broke around the nested scope>joins above.